core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -549,8 +558,17 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
         String sSource = "_na_";
         try {
-            sSource = XContentHelper.convertToJson(source, false);
-        } catch (Exception e) {
+            if (source.length() > MAX_SOURCE_LENGTH_IN_TOSTRING) {
+                // We might slice in the middle of a UTF-8 character here, but BytesRef.utf8ToString() will read more bytes if needed.
+                // There is no risk of ArrayIndexOutOfBoundsException either as source is assumed to be properly UTF-8 encoded.
+                sSource = source.slice(0, MAX_SOURCE_LENGTH_IN_TOSTRING).utf8ToString() + "..." +
```

Comment:
@sabi0 I was thinking that if you override writeBytes instead of writeByte and ignore entire content of the method it should work, but it's a moot point now because I think @bleskes's idea of not including source at all is much better.